hybrasyl/Objects/User.cs

```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
+                Client.Enqueue(new ServerPacketStructures.Cooldown()
                 {
-                    if (item != null)
-                    {
-                        if (!(Inventory.Contains(item.Value, item.Quantity))) return false;
-                    }
-                }
+                    Length = (uint)castable.Cooldown,
+                    Pane = 1,
+                    Slot = slot
+                }.Packet());
             }
-
-            return true;
+            else
+                SendSystemMessage("It didn't work.");
         }
 
-        public void CastingCost(Castable castable)
+        /// <summary>
+        /// Process the casting cost for a castable. If all requirements were not met, return false.
+        /// </summary>
+        /// <param name="castable">The castable that is being cast.</param>
+        /// <returns>True or false depending on success.</returns>
+        public bool ProcessCastingCost(Castable castable)
         {
             var costs = castable.CastCosts.CastCost;
             uint reduceHp = 0;
             uint reduceMp = 0;
             int removeNumItems = 0;
 
-            if (double.TryParse(costs.Stat.Hp, out double hpCostPercent))
-            {
-                if (hpCostPercent != 0)
-                {
-                    reduceHp = (uint)Math.Ceiling(this.MaximumHp * hpCostPercent);
-                }
-            }
+            // HP cost can be either a percentage (0.25) or a fixed amount (50)
+            if (costs.Stat.Hp.Contains('.'))
+                reduceHp = (uint) Math.Ceiling(Convert.ToDouble(costs.Stat.Hp) * this.MaximumHp);
+            else 
+                reduceHp = Convert.ToUInt32(costs.Stat.Hp);
 
-            if (double.TryParse(costs.Stat.Mp, out double mpCostPercent))
-            {
-                if (mpCostPercent != 0)
-                {
-                    reduceMp = (uint)Math.Ceiling(this.MaximumMp * mpCostPercent);
-                }
-            }
+            if (costs.Stat.Mp.Contains('.'))
+                reduceMp = (uint)Math.Ceiling(Convert.ToDouble(costs.Stat.Mp) * this.MaximumMp);
+            else
+                reduceMp = Convert.ToUInt32(costs.Stat.Mp);
 
+            
             if (costs.Items != null)
             {
                 foreach (var item in costs.Items)
                 {
-                    if (item != null)
-                    {
-                        if (Inventory.Contains(item.Value, item.Quantity)) removeNumItems++;
-                    }
+                    if (Inventory.Contains(item.Value, item.Quantity)) removeNumItems++;
                 }
             }
 
+            // Check that all requirements are met first. Note that a spell cannot be cast if its HP cost would result
+            // in the caster's HP being reduced to zero.
+            if (reduceHp > Hp || reduceMp >= Mp || costs.Gold > Gold || removeNumItems != costs.Items.Count) return false;

```

Comment:
Need null check for Items, unless you want to make this required element, with a default count of 0.

Reply:
Fixed.